### ISSUE TITLE
[Feature Request] [aptos-stdlib] support keccak hash with state

### ISSUE THREAD

**Issue by nanne007** [aptos-labs/aptos-core#11018]
# 🚀 Feature Request

Currently, aptos supports getting the hash of some bytes directly by calling  `aptos_hash::keccak256(vector<u8>)`.
it's very handy when hashing  small pieces of data.
However, when developers want to hash a lot of data, they need to concat the data into on big bytes, which is gas-consuming.
Besides, if developers want to continuously get hashes of an append-only bytes, they have to keep the size-growing bytes, and hash it repeatedly. 
A better solution is exposing `KeccakState`. developers only need to keep the state(it can be small if implemented natively), and update bytes data into the state, and get hash from state. a possible api for move is:

``` move
module halo2_verifier::keccak256 {
    #[test_only]
    use std::vector;

    native struct KeccakState;

    /// create a new keccak state
    native fun new(): KeccakState;
    /// update bytes into the keccak state
    native fun update(self: &mut KeccakState, bytes: &vector<u8>);
    /// get keccak hash from state
    native fun finalize(self: &KeccakState): vector<u8>;
    
    #[test]
    public fun test() {
        let bytes = b"1234";
        let i = 0;
        let hashes = vector::empty();
        let keccak = new();
        while (i < 100) {
            update(&mut keccak, &bytes);
            vector::push_back(&mut hashes, finalize(&keccak));
            i = i+1;
        };
    }
}

```

## Motivation

**Is your feature request related to a problem? Please describe.**

When [implementing transcript](https://github.com/young-rocks/halo2-verifier.move/pull/14) for our halo2-verifier, after appending some bytes into a keccak state, we usually need to get a hash from it. the case is similar to the test case above.

## Pitch

**Describe the solution you'd like**
Abstracting a common hash state, so that we can generalize the needs to other hash methods, like `blake2b_256`, `sha3_512`

**Describe alternatives you've considered**

**Are you willing to open a pull request?** (See [CONTRIBUTING](../../CONTRIBUTING.md))
Yes.

**Comment by nanne007** [comment]
Another approach, use plain move struct to store data needed to do keccak hashing, and make heavy-computation functions  like `update` and `finalize` native.

**Comment by zjma** [comment]
> Another approach, use plain move struct to store data needed to do keccak hashing, and make heavy-computation functions like `update` and `finalize` native.

I think this approach is better if there is a rust library that supports hash state (de-)serialization. @nanne007 do you know one?

**Comment by nanne007** [comment]
> > Another approach, use plain move struct to store data needed to do keccak hashing, and make heavy-computation functions like `update` and `finalize` native.
> 
> I think this approach is better if there is a rust library that supports hash state (de-)serialization. @nanne007 do you know one?

maybe we don't need serialization?

A posssible impl is :

``` move
module halo2_verifier::keccak {
   /// copied from tiny-keccak in rust.
    struct KeccakState {
        buffer: vector<u64>, // of length 25
        offset: u32,
        rate: u32,
        delim: u8,
        mode: bool,
    }
    
    public fun new(): KeccakState {
        // todo
        abort 1
    }
    public native fun update(self: &mut KeccakState, data: &vector<u8>);
    public native fun finalize(self: &KeccakState): vector<u8>;
}
```
when implementing the native methods, we just need to pop this plain move struct.
no serialization needed.

**Comment by zjma** [comment]
right, we want the hash state in rust somehow manually recoverable, either by deserialization, or by low-level pieces in rust.

I think that's not supported in our current dependencies: sha3 or tiny-keccak.

Let's try making a change to those. sha3 looks more promising.

**Comment by nanne007** [comment]
> right, we want the hash state in rust somehow manually recoverable, either by deserialization, or by low-level pieces in rust.
> 
> I think that's not supported in our current dependencies: sha3 or tiny-keccak.
> 
> Let's try making a change to those. sha3 looks more promising.

Yes, changes need to be made to make it possible.

And another issue we should consider is that other hash functions like blake2b_256, sha3_512 share the same problem here. If we try to generalize the solution to them. A possible api is:

``` move
module std::hash_internal {

    public(friend) native fun update<S>(self: &mut S, data: &vector<u8>);

    public(friend) native fun finalize<S>(self: &S): vector<u8>;

}

module std::keccak {
    use std::hash_internal;

    /// copied from tiny-keccak in rust.
    struct KeccakState {
        buffer: vector<u64>,
        // of length 25
        offset: u32,
        rate: u32,
        delim: u8,
        mode: bool,
    }

    public fun new(): KeccakState {
        // todo
        abort 1
    }

    public fun update(self: &mut KeccakState, data: &vector<u8>) {
        hash_internal::update(self, data)
    }
    public fun finalize(self: &KeccakState): vector<u8> {
        hash_internal::finalize(self)
    }
}

module std::blake2b {
    use std::hash_internal;

    struct Blake2bState {
        // todo
    }

    public fun new(): Blake2bState {
        // todo
        abort 1
    }

    public fun update(self: &mut Blake2bState, data: &vector<u8>) {
        hash_internal::update(self, data)
    }
    public fun finalize(self: &Blake2bState): vector<u8> {
        hash_internal::finalize(self)
    }
}
```

so that we can use the  two native functions to support all the hash methods.

**Comment by luxebeng** [comment]
There is a Keccak hash function available by pure Move. https://github.com/luxebeng/aptos_keccak
And gas cost evaluation is made between pure Move function and Native function. (please refer to source code for detail).
It's found that native hash function is about ~270 times gas reduction.(39 vs 10488).

| Test Case | Function Executed | Gas Used |
| :---: | :---: | :---: |
| case 4 | new + 10 * (update + finalize) | 10488 |
| case 5 |  10 * native hash function | 39 |
